### DISCUSSION TITLE
TreeGridMixin in Serenity.EntityGrid with MasterDetailRelation

### DISCUSSION THREAD

**Discussion by clarklin7475** [serenity-is/Serenity#6132] [Help]
![1638870667583](https://user-images.githubusercontent.com/24400518/145006658-fcfc35ae-050b-4f3a-9438-57583c23c077.jpg)

my tree is detail, but it was not  filtered by SystemId
SystemId is foreignKey.

My editor Code:

namespace SystemMenuGroupAuthority.SystemDB {

    @Serenity.Decorators.registerEditor()
    export class SystemMenuTreeGrid extends Serenity.EntityGrid<SystemMenuRow, any> {//Extensions.GridEditorBase   EntityGrid
        protected getColumnsKey() { return 'SystemDB.SystemMenu'; }
        protected getDialogType() { return SystemMenuDialog; }
        protected getIdProperty() { return SystemMenuRow.idProperty; }
        //protected getInsertPermission() { return SystemMenuRow.insertPermission; }
        //protected getLocalTextPrefix() { return SystemMenuRow.localTextPrefix; }
        protected getService() { return SystemMenuService.baseUrl; }
        private treeMixin: Serenity.TreeGridMixin<SystemDB.SystemMenuRow>;
        //constructor(container: JQuery) {
        //    super(container);
        //}
        public SystemId: number;

        constructor(container: JQuery) {
            super(container);
            this.treeMixin = new Serenity.TreeGridMixin({
                grid: this,
      
                // bring tree items initially collapsed
                initialCollapse: () => true,
                // which column to place tree toggle / expand/collapse button
                toggleField: SystemDB.SystemMenuRow.Fields.Name,
                getParentId: x => {
                  
                    return x.ParentCode;
                }

            });
        }

        protected getAddButtonCaption() {
            return "新增";
        }
    
    }
}

Row.cs:
      [MasterDetailRelation(foreignKey: "SystemId")]//, IncludeColumns = "FileName"
        [DisplayName("MenuList"), NotMapped]
        public List<SystemMenuRow> SystemMenuList
        {
            get => fields.SystemMenuList[this];
            set => fields.SystemMenuList[this] = value;
        }

where is wrong?

**Comment by wezmag** [comment]
Your detail tree was not filtered by SystemId is because your editor extends Serenity.EntityGrid. 
The data was retrieved from SystemMenuService.List without applying any equality filter.

Normally when using MasterDetailRelation, we extend editor class from Extensions.GridEditorBase.
Not sure why you commented out and use Serenity.EntityGrid instead?

**Comment by clarklin7475** [comment]
if i use Extensions.GridEditorBase, my tree grid was wrong

![1638927400634](https://user-images.githubusercontent.com/24400518/145133301-7e7723c3-eb9d-47ea-bb9a-72367fa6c858.jpg)
![1638927611354](https://user-images.githubusercontent.com/24400518/145133304-4a996acc-50ff-45a0-8efe-79ed1e51066f.jpg)

**Reply by wezmag** [reply]
Please refer to https://github.com/serenity-is/Serenity/issues/4885 and https://github.com/serenity-is/Serenity/issues/3152 .
I think you will need to override the `onViewProcessData` event. https://github.com/serenity-is/Serenity/issues/4885#issuecomment-595654169
